go/pkg/services/log_collector/server.go

```diff
@@ -16,36 +16,507 @@ package log_collector
+	abstractServer, err := framework.NewAbstractMlrunGRPCServer(logger, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to create abstract server")
+	}
+
+	// initialize kubernetes client
+	restConfig, err := common.GetKubernetesClientConfig(kubeconfigPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to get client configuration")
+	}
+	kubeClientSet, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to create kubernetes client set")
+	}
+
+	// parse interval durations
+	stateFileUpdateIntervalDuration, err := time.ParseDuration(stateFileUpdateInterval)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to parse monitoring interval")
+	}
+
+	// ensure base dir exists
+	if err := common.EnsureDirExists(baseDir, os.ModeDir); err != nil {
+		return nil, errors.Wrap(err, "Failed to ensure base dir exists")
+	}
+
+	return &LogCollectorServer{
+		AbstractMlrunGRPCServer: abstractServer,
+		namespace:               namespace,
+		baseDir:                 baseDir,
+		stateFileUpdateInterval: stateFileUpdateIntervalDuration,
+		kubeClientSet:           kubeClientSet,
+		stateLock:               &sync.Mutex{},
+		stateFileLock:           &sync.Mutex{},
+	}, nil
 }
 
-func (s *logCollectorServer) StartLog(ctx context.Context, request *log_collector.StartLogRequest) (*log_collector.StartLogResponse, error) {
+func (lcs *LogCollectorServer) OnBeforeStart(ctx context.Context) error {
+	lcs.Logger.DebugCtx(ctx, "Initializing Server")
+
+	// load state from file
+	state, err := lcs.readStateFile()
+	if err != nil {
+		return errors.Wrap(err, "Failed to load state file")
+	}
+	lcs.state = state
+
+	// start state updating goroutine
+	go lcs.updateStateFile(ctx)
+
+	// if there are items in the stateFile, call StartLog for each of them
+	for runId, logItem := range state.InProgress {
+		lcs.Logger.DebugWithCtx(ctx, "Starting log collection for log item", "runId", runId)
+		if _, err := lcs.StartLog(ctx, &log_collector.StartLogRequest{
+			RunId:    runId,
+			Selector: logItem.LabelSelector,
+		}); err != nil {
+			lcs.Logger.WarnWithCtx(ctx, "Failed to start log collection for log item", "runId", runId)
+			continue
+		}
+	}
+
+	return nil
+}
+
+func (lcs *LogCollectorServer) RegisterRoutes(ctx context.Context) {
+	lcs.AbstractMlrunGRPCServer.RegisterRoutes(ctx)
+	log_collector.RegisterLogCollectorServer(lcs.Server, lcs)
+}
+
+// StartLog writes the log item info to the state file, gets the pod using the label selector,
+// triggers `monitorPod` and `streamLogs` goroutines.
+func (lcs *LogCollectorServer) StartLog(ctx context.Context, request *log_collector.StartLogRequest) (*log_collector.StartLogResponse, error) {
+
+	lcs.Logger.DebugWithCtx(ctx, "Received Start Log request", "request", request)
+
+	// write log item in progress to file
+	if err := lcs.addItemToInProgress(ctx, request.RunId, request.Selector); err != nil {
```

Comment:
Id expect the state file to be an interface that expose functionality that the logcollector will use, e.g.:
`fileStateStore.add(...)`, `fileStateStore.remove()`, etc